sktime/utils/_testing/deep_equals.py

```diff
@@ -30,52 +32,93 @@ def deep_equals(x, y):
     """
+
+    def ret(is_equal, msg):
+        if return_msg:
+            if is_equal:
+                msg = ""
+            return is_equal, msg
+        else:
+            return is_equal
+
     if type(x) != type(y):
-        return False
+        return ret(False, f".type, x.type = {type(x)} != y.type = {type(y)}")
 
     # we now know all types are the same
     # so now we compare values
     if isinstance(x, pd.Series):
         if x.dtype != y.dtype:
-            return False
+            return ret(False, f".dtype, x.dtype= {x.dtype} != y.dtype = {y.dtype}")
         # if columns are object, recurse over entries and index
         if x.dtype == "object":
             index_equal = x.index.equals(y.index)
-            return index_equal and deep_equals(list(x.values), list(y.values))
+            values_equal, values_msg = deep_equals(
+                list(x.values), list(y.values), return_msg=True
+            )
+            if not values_equal:
+                msg = ".values" + values_msg
+            elif not index_equal:
+                msg = f".index, x.index: {x.index}, y.index: {y.index}"
+            else:
+                msg = ""
+            return ret(index_equal and values_equal, msg)
         else:
-            return x.equals(y)
+            return ret(x.equals(y), f".series_equals, x = {x} != y = {y}")
     elif isinstance(x, pd.DataFrame):
         if not x.columns.equals(y.columns):
-            return False
+            return ret(
+                False, f".columns, x.columns = {x.columns} != y.columns = {y.columns}"
+            )
         # if columns are equal and at least one is object, recurse over Series
         if sum(x.dtypes == "object") > 0:
-            return np.all([deep_equals(x[c], y[c]) for c in x.columns])
+            for c in x.columns:
+                is_equal, msg = deep_equals(x[c], y[c], return_msg=True)
+                if not is_equal:
+                    return ret(False, f'["{c}"]' + msg)
+            return ret(True, "")
         else:
-            return x.equals(y)
+            return ret(x.equals(y), f".df_equals, x = {x} != y = {y}")
     elif isinstance(x, np.ndarray):
         if x.dtype != y.dtype:
-            return False
-        return np.array_equal(x, y, equal_nan=True)
+            return ret(False, f".dtype, x.dtype = {x.dtype} != y.dtype = {y.dtype}")
+        return ret(np.array_equal(x, y, equal_nan=True), ".values")
     # recursion through lists, tuples and dicts
     elif isinstance(x, (list, tuple)):
-        return _tuple_equals(x, y)
+        return ret(*_tuple_equals(x, y, return_msg=True))
     elif isinstance(x, dict):
-        return _dict_equals(x, y)
+        return ret(*_dict_equals(x, y, return_msg=True))
     elif x != y:
-        return False
+        return ret(False, f" !=, {x} != {y}")
 
-    return True
+    return ret(True, "")
 
 
-def _tuple_equals(x, y):
+def _tuple_equals(x, y, return_msg=False):
```

Comment:
Have you tried https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertTupleEqual instead?

Reply:
same, wouldn´t deal with the recursion, since we need to call `deep_equals` from `_dict_equals`